init dev (#1)
AlexanderBrockmann/swt2_hausarbeit: Latex/samplepaper.tex

% This is samplepaper.tex, a sample chapter demonstrating the
% LLNCS macro package for Springer Computer Science proceedings;
% Version 2.20 of 2017/10/04
%
\documentclass[runningheads]{llncs}
\usepackage{listings}
%
\usepackage{graphicx}
% Used for displaying a sample figure. If possible, figure files should
% be included in EPS format.
%
% If you use the hyperref package, please uncomment the following line
% to display URLs in blue roman font according to Springer's eBook style:
% \renewcommand\UrlFont{\color{blue}\rmfamily}

\usepackage{german}
\usepackage[utf8]{inputenc}

\begin{document}
%
\title{SmartShop}
%
%\titlerunning{Abbreviated paper title}
% If the paper title is too long for the running head, you can set
% an abbreviated paper title here
%
\author{Alexander Brockmann}
%
% First names are abbreviated in the running head.
% If there are more than two authors, 'et al.' is used.
%
\institute{7202423\\
\email{{alexander.brockmann002}@stud.fh-dortmund.de}}

%
\maketitle              % typeset the header of the contribution

\section{Beschreibung der Projektidee}

> Durch Covid-19 ist es in vielen Bereichen des täglichen Lebens aufgefallen, wie wichtig es ist Menschen zu koordinieren, beziehungsweise Menschenmassen zu vermeiden.

Reviewer: ```suggestion
Durch Covid-19 ist es in vielen Bereichen des täglichen Lebens aufgefallen, wie wichtig es ist, Menschen zu koordinieren beziehungsweise große Menschenansammlungen zu vermeiden.
```

Auch außerhalb von Pandemien lässt sich die bestehende Infrastruktur durch eine gleichmäßige Verteilung bestmöglich nutzen.
Auf diesen Erkenntnissen basierend entstand die Idee einer verteilten Anwendung zur Optimierung der Auslastung von Supermärkten.

> Die Anwendung soll dem Anwender sagen, wie ausgelastet ein Supermarkt ist, beziehungsweise welcher Supermarkt logistisch und auslastungstechnisch am besten gelegen ist.

Reviewer: ```suggestion
Die Anwendung soll dem Anwender Informationen zur aktuellen Auslastung eines Supermarktes geben.
Außerdem sollen dem Anwender Supermärkte anhand der Auslastung und der erreichbarkeit empfohlen werden.
```

> Zu dem System gehören die Ein- und Ausgangssensoren von Supermärkten, Standortdaten von Nutzern, eine Forecast AI zum prognostizieren von möglichen Auslastungen, eine Datenbank zum Vorhalten von sensorischen Daten, einem Frontend Client und einem Backend zum Aufbereiten und Bereitstellen von Daten.

Reviewer: ```suggestion
Zu dem System gehören die Ein- und Ausgangssensoren von Supermärkten, Standortdaten von Nutzern, eine Forecast AI zum prognostizieren von möglichen Auslastungen, eine Datenbank zum Vorhalten von sensorischen Daten, ein Frontend Client zur Bedienung durch den Nutzer und ein Backend zum Aufbereiten und Bereitstellen der Daten.
```

Durch die Kombination von statistischen Erfahrungswerten mit Umweltdaten wie dem Wetter, stattfindenden Events oder Baustellen und Streckensperrungen soll bestimmt werden, wann welcher Supermarkt am besten zu nutzen ist.
Darüber hinaus sollen Empfehlungen zur Route beziehungsweise den zu nutzenden öffentlichen Verkehrsmitteln gegeben werden.

> Zur Anwendungsdomäne gehören folgende Entitäten.

Reviewer: ```suggestion
Zur Anwendungsdomäne gehören folgende Entitäten:
```

\begin{description}

> 	\item[Sensordaten] Ein Sensor erfasst seinen aktuellen Standort, die aktuelle Zeit, die Kapazität seines zugeordneten Supermarktes sowie dessen momentane Auslastung. Ein solcher Sensor ist mittlerweile in vielen Supermärkten im Einsatz. Die oben genannten Attributen werden in der Entität \textit{SensorData} zusammengefasst und in regelmäßigen Abständen bereit gestellt.

Reviewer: Den Text hier musst du anpassen, der Sensor hat keine Informationen über die Kapazität des Supermarktes

```suggestion
	\item[Sensordaten] Ein Sensor erfasst seinen aktuellen Standort, die aktuelle Zeit, die Kapazität seines zugeordneten Supermarktes sowie dessen momentane Auslastung. Ein solcher Sensor ist mittlerweile in vielen Supermärkten im Einsatz. Die oben genannten Attribute werden in der Entität \textit{SensorData} zusammengefasst und in regelmäßigen Abständen bereitgestellt.
```

> 	\item[Nutzer] Ein Nutzer hat einen Standort, welcher von dem Client an das System bei der Anfrage weiter gegeben wird. Zusätzlich wählt der Nutzer aus, nach welcher Art Shop er suchen möchte. Auf dieser Basis bekommt er einen Shop und gegebenenfalls eine Route.

Reviewer: ```suggestion
	\item[Nutzer] Ein Nutzer hat einen Standort, welcher vom Client bei der Anfrage an das System weitergegeben wird. Zusätzlich wählt der Nutzer aus, nach welcher Art Shop er suchen möchte. Auf dieser Basis bekommt er einen Shop und gegebenenfalls eine Route empfohlen.
```

	\item[GeoPosition] Die GeoPosition gibt die geografische Position basierend auf dem Längen- und Breitengrad an.

> 	\item[Shop] Der Shop repräsentiert ein Geschäft oder eine Ansammlung von Geschäften beispielsweise ein Einkaufszentrum. Zu einem Shop gehört jeweils ein Name, ein ShopType, die Größe der Ladenfläche, die Kapazität und die GeoPosition. Handelt es sich beispielsweise um ein Einkaufszentrum gehören dazu noch eine Liste weiterer Shops.

Reviewer: Zusätzlich hast du noch die Klasse SensorArea, die solltest du hiervor beschreiben.

```suggestion
	\item[Shop] Der Shop repräsentiert ein Geschäft oder eine Ansammlung von Geschäften, beispielsweise ein Einkaufszentrum. Zu einem Shop gehört jeweils ein Name, ein ShopType, die Größe der Ladenfläche, die Kapazität und die GeoPosition. Zusätzlich wird noch eine Liste an SensorAreas gespeichert.
```

> 	\item[ShopType] Der ShopType gibt an um was für eine Art Shop es sich handelt.

Reviewer: ```suggestion
	\item[ShopType] Der ShopType gibt an, um was für eine Art Shop es sich handelt.
```

Abhängig von der Art des Shops und seiner Größe wird auch die Kapazität berechnet.
	
\end{description}

\section{Architektur des Projektes}

\includegraphics[width=\linewidth]{images/deployment_diagram}

> \subsection{Beschreibung der Systemarchitektur}

Reviewer: In diesem Absatz beschreibst du, was in dem Bild zu sehen ist. Anstatt einfach nur das Bild zu beschreiben solltest du noch darauf eingehen, warum du dich für genau diese Architektur entschieden hast. Bspw:
* Warum Ubuntu und nicht Windows Server oder CentOS
* Warum diese Hardware und nicht andere
* Warum eine Java-Anwendung und nicht .NET oder NodeJS
* Warum MariaDB und nicht OracleDB, MongoDB oder ähnliches
* Warum Web-Frontend anstelle von Java Swing oder Java-FX

Author: Ubuntu ist kostenlos, nicht CentOS da nicht weiterentwickelt
Python, weil gängige Sprache für Ki, Python 3.9 aktuelle stabile
Java weil meiste Erfahrung und Java 11 weil LTS
MariaDB weil OpenSource und kostenlos
Embedded Tomcat weil Spring

Des Weiteren liefert das Backend den Frontend-Client als Thymeleaf-Anwendung aus

Vlt in nem Nebensatz, dass ich mich an den Vorlesungsbeispielen orientiert habe?

> Der Server auf dem die Anwendung läuft ist ein Sun Ice 9001 Server.

Reviewer: ```suggestion
Der Server, auf dem die Anwendung läuft, ist ein Sun Ice 9001 Server.
```

Dieser wird mit dem Betriebssystem Ubuntu 20.04 LTS betrieben.

> Darin wird zum einen die SmartAi.py ausgeführt, welche in der Laufzeitumgebung(runtime environment) Python 3.5 läuft.

Reviewer: ```suggestion
Darin wird zum einen die SmartAi.py ausgeführt, welche in der Laufzeitumgebung Python 3.5 läuft.
```

Zum anderen wird dort die SmartShopping.jar in der Laufzeitumgebung Java 14 ausgeführt.
SmartShopping.jar stellt einen Embedded Tomcat bereit.

> Die beiden Anwendungen kommunizieren intern via Rest-Schnittellen mit dem HTTP Protokoll.

Reviewer: ```suggestion
Die beiden Anwendungen kommunizieren intern via Rest-Schnittellen mit dem HTTP-Protokoll.
```

> Mit dem Datenbankserver kommunizieren sie mit dem JDBC Protokoll.

Reviewer: ```suggestion
Mit dem Datenbankserver kommunizieren sie mit dem JDBC-Protokoll.
```

Der Datenbankserver ist ein IBM z15.
Auf diesem läuft ein MariaDB 10.6.4.

> Dieser hat die Schemas SensorData und Shop.

Reviewer: ```suggestion
In der Datenbank gibt es die Schematas SensorData und Shop.
```

> Des Weiteren liefert das Backend den Frontend Client aus.

Reviewer: Das Backend liefert nicht den Frontend Client aus, das passt nicht. Je nach Art deines Clients wird der Client bei einer Web-Anwendung auf einem Server (ggf. deinem Sun Ice) gehostet oder auf dem Client-Gerät installiert, wenn du eine Java-Applikation bspw. mit Swing oder FX hast.

Reviewer: Unten schreibst du, du hast eine Webseite, die über Chrome aufgerufen wird, du müsstest dann die Bereitstellung davon noch modellieren.

> Welcher beispielsweise auf einem Chromebook ausgeliefert werden kann.

Reviewer: Je nach dem wie du es oben schreibst, musst du diesen Satz auch noch inhaltlich anpassen.

```suggestion
Der Frontend Client kann dann beispielsweise auf einem Chromebook ausgeliefert werden.
```

> Das Frontend kann dann zum Beispiel in Google Chrome aufgerufen werden.

Reviewer: ```suggestion
Das Frontend kann dann in einem Brwoser, zum Beispiel in Google Chrome, aufgerufen werden.
```

> Backend und Frontend kommunizieren mit dem HTTPS Protokoll.

Reviewer: ```suggestion
Backend und Frontend kommunizieren über das HTTPS-Protokoll.
```

> Außerdem wird im Verteilungsdiagramm ein MqttBroker und die dazu gehörenden Sensoren dargestellt.

Reviewer: ```suggestion
Außerdem werden im Verteilungsdiagramm ein MqttBroker und die dazu gehörenden Sensoren dargestellt.
```

> Da der Fokus nicht auf dem erheben der Daten liegt, wird nicht näher auf den MqttBroker und die dazu gehörenden Sensoren eingegangen.

Reviewer: ```suggestion
Da der Fokus nicht auf dem Erheben der Daten liegt, wird nicht näher auf den MqttBroker und die dazu gehörenden Sensoren eingegangen.
```

\subsection{Einordnung der Architekturentscheidungen}

> Der Architekturstil nach Starke ist ein Heterogenes System, da die Kommunikation zwischen den Services über Rest Schnittstellen realisiert ist.

Reviewer: Du hast hier eine Referenz auf eine Quelle, da brauchst du auch eine Zitierungsmarke.

```suggestion
Der Architekturstil nach Starke ist ein heterogenes System, da die Kommunikation zwischen den Services über REST-Schnittstellen realisiert ist.
```

Des Weiteren erfolgt die Kommunikation zwischen Anwendung via HTTP, HTTPS und JDBC.
Die Architektur wurde gewählt, um die Anwendung möglichst modular zu halten und so eine gute Austauschbarkeit von Komponenten zu gewährleisten.

\newpage
\section{Entwurfsmuster}

\subsection{Darstellung der Entwurfsmuster als Klassendiagramm}
\includegraphics[width=\linewidth]{images/OOD-Klassendiagram}
\begin{description}
\item[Komposit]

> Das Komposit ist so umgesetzt, sodass Komplexe Strukturen wie zum Beispiel Einkaufszentren dargestellt werden können.

Reviewer: ```suggestion
Das Komposit wird verwendet, damit komplexe Strukturen wie zum Beispiel Einkaufszentren dargestellt werden können.
```

> Dafür wird die abstrakte Klasse SensorBereich (SensorArea) genutzt. Diese besitzt die Methode zur Berechnung der Auslastung, dazu muss jeweils die Auslastung der darunter liegenden Läden(Shop) ebenfalls berechnet werden.

Reviewer: ```suggestion
Dafür wird die abstrakte Klasse SensorBereich (SensorArea) genutzt. Diese besitzt eine Methode zur Berechnung der Anzahl an Personen in der SensorArea, dazu muss jeweils die Anzahl an Personen in den untergeordneten SendorAreas ermittelt werden.
```

> Ein Laden besteht dabei aus mindestens einem Sensorbereich, wobei dieser aus weiteren Läden besteht oder aus konkreten Sensoren(Sensor).

Reviewer: ```suggestion
Eine SensorArea kann dabei ein Sensor oder ein Shop sein.
Handelt es sich um einen Shop, so verfügt dieser zusätzlich über eine Liste untergeordneter SensorAreas.
In einem normalen Supermarkt können diese den verschiedenen Eingängen des Supermarktes zugeordnet sein.
Bei einem Einkaufszentrum beispielsweise können die untergeordneten SensorAreas auch Shops sein.
```

> Dadurch kann der Sensorbereich beliebig skaliert und die eigentliche Auslastung rekursiv berechnet werden.  

Reviewer: ```suggestion
Dadurch kann die SensorArea beliebig skaliert und die eigentliche Anzahl an Personen in der SensorArea rekursiv berechnet werden.  
```

Diese Skalierbarkeit ist der Vorteil dieses Entwurfsmusters und wurde deswegen ausgewählt. 
\item[Beobachter-Muster]
Für das Beobachter-Muster wurde zunächst die Klasse MqttService modelliert. 

> Dieser nimmt Sensordaten vom Mqtt-Sensordaten-Server entgegen.

Reviewer: ```suggestion
Diese nimmt Sensordaten vom Mqtt-Sensordaten-Server entgegen.
```

> Damit diese Daten in das richtige Schema der Datenbank persistiert werden können, bietet der Service die Möglichkeit sich darauf einzutragen (subscribe).

Reviewer: ```suggestion
Damit diese Daten in das richtige Schema der Datenbank persistiert werden können, bietet der Service die Möglichkeit sich darauf zu subscriben.
```

> Eingetragene Beobachter (Observer) werden über eingehende Sensordaten informiert.

Reviewer: ```suggestion
Eingetragene Observer werden über eingehende Sensordaten informiert.
```

> Für die Beobachter wird eine Schnittstelle (Interface) bereit gestellt.

Reviewer: ```suggestion
Für die Observer wird ein Interface bereit gestellt.
```

Diese Schnittstelle wird von einer konkreten Klasse implementiert, dem SensorDataWorker.

> Dieser SensorDataWorker filtert die Daten, ruft den DatenbankService auf und übergibt die relevanten Daten.

Reviewer: ```suggestion
Dieser SensorDataWorker filtert die Daten, ruft den DatenbankService auf und übergibt die zu speichernden Daten.
```

> Durch dieses Beobachter-Muster ist es möglich weitere Daten über separate Beobachter zu verarbeiten, ohne das bisherige Beobachter angepasst werden müssen.

Reviewer: ```suggestion
Durch dieses Beobachter-Muster ist es möglich, weitere Daten über separate Beobachter zu verarbeiten, ohne das bisherige Beobachter angepasst werden müssen.
```

\end{description}

\subsection{Begründung}

> Der Vorteil des Kompositum-Musters ist, dass der Sensorenbereich (SensorArea) beliebig erweitert werden kann und es dabei zweitrangig ist um was die Shops beziehungsweise andere Komponenten erweitert werden. 

Reviewer: ```suggestion
Der Vorteil des Kompositum-Musters ist, dass die SensorArea beliebig erweitert werden kann und es dabei zweitrangig ist, um welche Attribute und Methoden die Shops beziehungsweise andere Komponenten erweitert werden. 
```

Ein Nachteil dabei ist, dass die Tiefe ebenfalls beliebig wächst und es so zu einer sehr unübersichtlichen Datenstruktur kommen kann.
Eine Alternative beziehungsweise Erweiterung wäre das Fabrikmethode-Muster. 
Dieses Muster erlaubt es uns auf Grundlage der abstrakten Klasse SensorArea weitere Bereiche zu implementieren, ohne dass davon andere Services betroffen sind. 

> Man könnte also Shop oder Einkaufszentrum erzeugen, würde dabei aber die Möglichkeit der Schachtelung verlieren. 

Reviewer: ```suggestion
Man könnte also einen Shop vom Typ Supermarkt oder vom Typ Einkaufszentrum erzeugen, würde dabei aber die Möglichkeit der Schachtelung verlieren. 
```

> Um beide Entwurfsmuster möglichst effektiv zu verwenden, wäre eine Kombination denkbar.

Reviewer: Warum nutzt du diese Kombination hier dann nicht, wenn sie doch die Probleme mit der tiefen Verschachtelung des Komposite behebt?

> \\
> 
> 

Reviewer: ```suggestion

```

Das Beobachter-Muster wurde gewählt, da es eingehende Signale von den Sensoren beziehungsweise Sensornachrichten automatisch verarbeitet und speichert.
Dieses Muster könnte durch das Zustands-Muster erweitert, aber nicht ersetzt werden.

> Das Zustands-Muster würde in der konkreten Implementierung ebenfalls auf eingehende Nachrichten warten / reagieren, es könnte beispielsweise als ein Buffer agieren welcher Nachrichten sammelt und erst auf Anfrage Daten weiter leitet / speichert.

Reviewer: ```suggestion
Das Zustands-Muster würde in der konkreten Implementierung ebenfalls auf eingehende Nachrichten warten beziehungsweise reagieren, es könnte beispielsweise als ein Buffer agieren welcher Nachrichten sammelt und erst auf Anfrage Daten weiterleitet oder speichert.
```

> Dieses Verhalten würde das Beobachter-Muster eher erweitern oder imitieren und ist daher ungeeignet in diesem Kontext.

Reviewer: Oben sagst du, das Beobachter-Muster könnte durch das Status-Muster erweitert werden, hier schließt du es aus genau den Gründen aus, warum ziehst du es dann überhaupt in Erwägung?

\newpage
\section{Beschreibung der Kernfunktionalität}

> Eine der Kernfunktionalitäten ist es, die Auslastung von Shops zu ermitteln und mit dieser Information weiter zu arbeiten.

Reviewer: ```suggestion
Eine der Kernfunktionalitäten der verteilten Anwendung ist es, die Auslastung von Shops zu ermitteln und mit diesen Informationen weiter zu arbeiten.
```

> Daher stellen wir in diesem Abschnitt die Funktionalität zur Ermittlung der Auslastung eines Einkaufzentrums (ShoppingCenter) dar.

Reviewer: ```suggestion
Daher stellen wir in diesem Abschnitt die Funktionalität zur Ermittlung der Auslastung eines Shops vom Typ ShoppingCenter dar.
```

\includegraphics[width=\linewidth]{images/Sequenzdiagramm}

> In dem Sequenzdiagramm beschriebenen Beispiel, möchte ein Client die Auslastung eines bestimmten Shops haben und ruft daher die Methode calculateUsage auf dem SmartShopService auf.

Reviewer: ```suggestion
In dem im Sequenzdiagramm beschriebenen Beispiel möchte ein Client die Auslastung eines bestimmten Shops abfragen und ruft daher die Methode calculateUsage auf dem SmartShopService auf.
```

\\

> \textbf{SmartShopService}: In diesem Beispiel wird davon ausgegangen, dass es bei der Auflösung einer bestimmten ShopOID nicht zu einem Fehler kommt sondern die Auflösung immer erfolgreich ist.

Reviewer: Den Effekt, den du hier versuchst zu erreichen, kannst du auch über \paragraph erreichen, das ist dann sogar noch etwas sauberer in latex.

> Die Methode \textbf{\textit{calculateUsage}} berechnet die aktuelle Auslastung des Shops welcher über seine ShopOID angefordert wird.

Reviewer: Weiter oben schreibst du Methodennamen wie den normalen Text, hier schreibst du sie fett und kursiv. Normalerweise reicht fett oder kursiv, wichtig ist aber, dass es in deiner ganzen Arbeit einheitlich ist.

```suggestion
Die Methode \textbf{\textit{calculateUsage}} berechnet die aktuelle Auslastung des Shops, welcher über seine ShopOID angefordert wird.
```

\begin{lstlisting}[language=Java, basicstyle=\scriptsize]
public int calculateUsage(int shopOID){
	Shop tmp = datenbankService.getShop(shopOID);

> 	return tmp.calculateUsage();

Reviewer: tmp ist kein guter Variablenname, nenn die Variable doch einfach shop.

}
\end{lstlisting}

\textbf{DatenbankService}: Beim Aufruf der Datenbank wird davon ausgegangen, dass es zu keinem Fehler kommt und die Datenbank den Shop mittels Primärschlüssel findet.
Die genaue Anfrage an die Datenbank wird durch die Funktion \textbf{\textit{getShopByPrimaryKey}} abstrahiert.
\begin{lstlisting}[language=Java, basicstyle=\scriptsize]
public Shop getShop(int shopOID){
	return getShopByPrimaryKey(shopOID);
}
\end{lstlisting}

\break

> \textbf{Shop}: Die Methode \textbf{\textit{calculateCustomerCount}} iteriert über die Liste seiner Sensoren und summiert deren Kunden auf.

Reviewer: ```suggestion
\textbf{Shop}: Die Methode \textbf{\textit{calculateCustomerCount}}des Shops iteriert über die Liste seiner Sensoren und summiert die von dem jeweiligen Sensor gezählten Kunden auf.
```

> Darauf hin iteriert die Methode über die Shops die dem aktuellen Shop zugeordnet sind und ruft darauf \textbf{\textit{calculateCustomerCount}} auf, um die Anzahl der zurück gegebenen Kunden aufzusummieren.

Reviewer: ```suggestion
Daraufhin iteriert die Methode über die Shops die dem aktuellen Shop zugeordnet sind und ruft darauf \textbf{\textit{calculateCustomerCount}} auf, um die Anzahl der zurückgegebenen Kunden aufzusummieren.
```

> \textbf{Shop}: Die Methode \textbf{\textit{calculateCustomerCount}} iteriert über die Liste seiner Sensoren und summiert deren Kunden auf.
> Darauf hin iteriert die Methode über die Shops die dem aktuellen Shop zugeordnet sind und ruft darauf \textbf{\textit{calculateCustomerCount}} auf, um die Anzahl der zurück gegebenen Kunden aufzusummieren.

Reviewer: ```suggestion
\textbf{Shop}: Zur Berechnung der Anzahl an Personen in der Methode \textbf{\textit{calculateCustomerCount}} des Shops wird über die Liste der SensorAreas iteriert.
Für jede SensorArea wird die wiederum die Anzahl an Personen berechnet und die Rückgabewerte werden im Shop aufsummiert.
```

> Die Methode liefert die aufsummierte Anzahl an Kunden pro Shop und Sensor zurück.

Reviewer: ```suggestion
Die Methode liefert die aufsummierte Anzahl an Kunden in dem Shop zurück.
```

\begin{lstlisting}[language=Java, basicstyle=\scriptsize]
public int calculateCustomerCount(){

> 	int customerCount = 0;
> 	List[Shop] shopList;
> 	List[Sensor] sensorList;
> 	for(Sensor sensor : sensorList){
> 		customerCount += sensor.getCustomerCount();
> 	}
> 	for(Shop shop : shopList){
> 		customerCount += shop.calculateCustomerCount();
> 	}
> 	return customerCount;
> }

Reviewer: Du hattest die Funktion calculateCustomerCount genannt. Außerdem funktioniert dein Code so nicht.
```suggestion
	int customerCount = 0;
	for (SensorArea sensorArea : this.getSensorAreas()) {
		customerCount += sensorArea.calculateCustomerCount();
	}
	return customerCount;
}
```

\end{lstlisting}

> Die Methode \textbf{\textit{calculateUsage}}, berechnet die momentane Auslastung eines Shops, basierend auf seiner Kapazität, der aktuellen Anzahl seiner Kunden und der Anzahl von Kunden seiner Shops.

Reviewer: ```suggestion
Die Methode \textbf{\textit{calculateUsage}}, berechnet die momentane Auslastung eines Shops basierend auf seiner Kapazität, der aktuellen Anzahl seiner Kunden und der Anzahl von Kunden seiner untergeordneten SensorAreas.
```

\begin{lstlisting}[language=Java, basicstyle=\scriptsize]
public int calculateUsage(){

> 	int customerCount = 0;
> 	List[Shop] shopList;
> 	List[Sensor] sensorList;
> 	for(Shop shop : shopList){
> 		customerCount = shop.calculateCustomerCount();
> 	}
> 	for(Sensor sensor : sensorList){
> 		customerCount = sensor.getCustomerCount();
> 	}
> 	return capacity - customerCount;

Reviewer: Du hattest die Funktion calculateCustomerCount genannt. Außerdem funktioniert dein Code so nicht. Und calculateUsage sollte einen relativen Wert zurückgeben, hattest du gesagt. Da passt dann deine Berechnung nicht.
```suggestion
	return this.calculateCustomerCount() / this.capacity;
```

}
\end{lstlisting}

\newpage
\section{Persistente Datenhaltung}

\subsection{Persistierungansatz}
\includegraphics[width=\linewidth]{images/Datamodel}
\begin{description}
	\item[Shop] Der Shop besitzt eine eindeutige ID, welche als Primärschlüssel dient. 

> 	Des weiteren die Felder: 

Reviewer: ```suggestion
	Des Weiteren besitzt er folgende Attribute: 
```

	\begin{description}

> 		\item [name:] Das Feld name repräsentiert den Namen des Shops.

Reviewer: ```suggestion
		\item [name:] Das Attribut name repräsentiert den Namen des Shops.
```

> 		\item [capacity:] Für die maximale Anzahl an Kunden im Shop.

Reviewer: ```suggestion
		\item [capacity:] Das Attribut capacity speichert die maximale Anzahl an Kunden, die sich in dem Shop aufhalten dürfen.
```

		\item [size:] Das Attribut size beschreibt die Grundfläche des Shops.

> 		\item [shopType:] ShopType repräsentiert den Typen des Shops.

Reviewer: ```suggestion
		\item [shopType:] ShopType repräsentiert den Typ des Shops.
```

> 		\item [lat, lng:] lat, lng beschreiben den Breitengrad, Längengrad des Shops. Breitengrad und Längengrad kommen aus der Entität \textbf{GeoPosition}, welche aber in diesem Kontext in Shop persistiert werden, da sie in einer eins zu eins Beziehung stehen.

Reviewer: ```suggestion
		\item [lat, lng:] lat und lng beschreiben den Breiten- und Längengrad des Shops. Die Daten kommen aus der Entität \textbf{GeoPosition}, welche aber in diesem Kontext innerhalb des Shops persistiert wird, da Shop und GeoPosition in einer eins-zu-eins-Beziehung stehen.
```

> 		\item [parentOID:] Repräsentiert einen Fremdschlüssel auf einen möglicherweise übergeordneten Shop, das Feld kann leer sein.

Reviewer: ```suggestion
		\item [parentOID:] Die parentOID ist ein Fremdschlüssel auf eine möglicherweise übergeordneten SensorArea, das Attribut kann leer sein.
```

	\end{description} 

> 	Die Parent-Child-Beziehung beschreibt, das im Entwurfsmuster dargestellte Komposit.

Reviewer: ```suggestion
	Die Parent-Child-Beziehung beschreibt das im Entwurfsmuster dargestellte Komposit.
```

> 	\item[Sensor] Der Sensor besitzt das Feld sensorOID, welches eine eindeutige ID ist und als Primärschlüssel fungiert. Außerdem besitzt es das Feld shopOID als Fremdschlüssel, welcher auf einen eindeutigen Shop zeigt, das Feld kann leer sein.

Reviewer: ```suggestion
	\item[Sensor] Der Sensor besitzt das Feld sensorOID, welches eine eindeutige ID ist und als Primärschlüssel fungiert. Außerdem besitzt er das Attribut shopOID als Fremdschlüssel, welches auf einen Shop verweist, das Attribut kann leer sein.
```
Warum kann shopOID leer sein? Wie kann kann den ein Sensor ohne entsprechenden Shop existieren?

Author: Die shopOID kann leer sein, da ein Sensor nicht zwangsläufig einen Shop braucht. Ein Sensor ohne Shop ist zwar Quatsch aber möglich.

> 	\item[SensorData] SensorData besitzt die Felder:

Reviewer: ```suggestion
	\item[SensorData] SensorData besitzt folgende Attribute:
```

	\begin{description}

> 		\item[dataOID:] Die dataOID fungiert als ID und einzigartiger Primärschlüssel.

Reviewer: ```suggestion
		\item[dataOID:] Die dataOID ist eine eindeutige ID und fungiert als Primärschlüssel.
```

> 		\item[customerCount:] Das Feld customerCount repräsentiert die Anzahl an Kunden zu dem Zeitpunkt des Erstellens von SensorData.

Reviewer: ```suggestion
		\item[customerCount:] Das Attribut customerCount repräsentiert die Anzahl an Kunden innerhalb der zugeordneten SensorArea zu dem Zeitpunkt des Erstellens des SensorData-Objektes.
```

> 		\item[timeStamp:] Dieses Feld enthält den Zeitpunkt zu dem SensorData erstellt wurde.

Reviewer: ```suggestion
		\item[timeStamp:] Dieses Attribut speichert den Zeitpunkt, zu dem das SensorData-Objekt erstellt wurde.
```

		\item[sensorOID] SensorOID enthält einen Fremdschlüssel, welcher den zu SensorData gehörenden Sensor identifiziert.
	\end{description}
\end{description}

> In der Datenhaltung werden die Entitäten Shop, Sensor und Sensordaten (SensorData) gespeichert.

Reviewer: ```suggestion
In der Datenhaltung werden die Entitäten Shop, Sensor und SensorData gespeichert.
```

> Eine weitere Entität wäre die geografische Position, da diese aber in einer eins zu eins Beziehung zur Entität Shop steht, werden die Daten direkt in Shop integriert.

Reviewer: ```suggestion
Eine weitere Entität wäre die geografische Position, da diese aber in einer eins-zu-eins-Beziehung zur Entität Shop steht, werden die Daten direkt im Shop gespeichert.
```

> Der DatenbankService verwendet JPA und setzt dabei auf einen Entity Manager.

Reviewer: ```suggestion
Der DatenbankService verwendet Spring Data JPA.
Dadurch werden sowohl die benötigten Tabellen als auch die via Methodennamen definierten Datenbankabfragen automatisch generiert.
```

Der Entity Manager wird unter der Haube von Spring verwendet, das ist für deine Arbeit eigentlich nicht wichtig. Stattdessen kannst du noch etwas näher auf die Vorteile von Spring Data JPA eingehen, bspw. der gesparte Entwicklungsaufwand oder die Konfiguration der Datenbankverbindung, die dir abgenommen wird.

Die gewählte Datenbank ist eine MariaDB, welche auf einem eigenen Server ausgeliefert wird.

\subsection{Begründung}
Da die Daten einfach strukturiert sind und eine eindeutige ID besitzen, habe ich mich für eine SQL Datenbank entschieden.
Durch diese Datenstruktur ist ein schneller, sicherer und einfacher Zugriff möglich.

> Die gewählte Datenbank ist eine MariaDB, da diese frei verfügbar ist (OpenSource) und von Michael Widenius, dem Entwickler von MySQL, entwickelt wurde.

Reviewer: ```suggestion
Die gewählte Datenbank ist eine MariaDB, da diese frei verfügbare OpenSource-Lösung zu MySQL ist.
```

> Die Datenbank wird auf einem eigens dafür vorgesehenen Server aufgesetzt, um eine lose Kopplung zu gewährleisten. Außerdem können Anwendungen, anders als bei einer In-Memory Datenbank, unabhängig voneinander auf die Datenbank zugreifen.

Reviewer: ```suggestion
Die Datenbank wird auf einem eigens dafür vorgesehenen Server aufgesetzt, um eine lose Koppelung zu gewährleisten. Außerdem können Anwendungen, anders als bei einer In-Memory Datenbank, unabhängig voneinander auf die Datenbank zugreifen.
```

> Ich habe mich für JPA entschieden, da es durch Spring standardisiert ist und datenbankunabhängig funktioniert. Der Nachteil an JPA ist, dass weitere Abhängigkeiten in den Code geladen werden.

Reviewer: ```suggestion
Ich habe mich für Spring Data JPA entschieden, da es durch Spring standardisiert ist und datenbankunabhängig funktioniert. Der Nachteil an Spring Data JPA ist, dass weitere Abhängigkeiten in den Code geladen werden.
```

> Eine Alternative zu MariaDB wäre eine NoSQL Datenbank, ich habe mich dagegen entschieden, da ich mit einem festen Schema arbeiten möchte und NoSQL das nicht bietet.

Reviewer: ```suggestion
Eine Alternative zu MariaDB wäre neben dem bereits erwähnten MySQL eine dokumentenorientierte Datenbank.
Ich habe mich gegen diese Form von Datenbanken entschieden, da ich mit einem festen Schema arbeiten möchte und NoSQL dies nicht bietet.
```

> Für eine NoSQL Datenbank würde sprechen, dass man die Sensordaten welche mit einem Zeitstempel versehen werden in eine Zeitreihen-Datenbank auslagern kann.

Reviewer: ```suggestion
Da die Sensordaten mit einem Zeitstempel versehen werden, könnten sie in einer Zeitreihen-Datenbank gespeichert werden.
```
Warum machst du das dann nicht?

> Weitere Datenbankenmodelle wären zum Beispiel eine dokumentenorientierte Datenbank oder Graphdatenbank. Beide Modelle eignen sich für diesen Anwendungsfall nicht, da weder mit Dokumenten noch mit Graphen gearbeitet wird.

Reviewer: Dokumenten- und Graphdatenbanken fallen unter den Begriff NoSQL-Datenbanken. 

```suggestion
Graphdatenbank eignen sich für diesen Anwendungsfall nicht, da die zu speichernden Daten nicht in einem Graph dargestellt werden können.
```

\newpage

\section{Kommunikation}

\subsection{Kommunikationsansatz}
\includegraphics[width=\linewidth]{images/Kommunikation}

Die Kommunikation zwischen SmartAI und dem AiService erfolgt über REST-Schnittstellen mit dem HTTP-Protokoll, diese Kommunikation ist synchron.

> Die SmartAi kommuniziert mit dem Datenbankserver via JDBC, diese Kommunikation ist ebenfalls synchron. JDBC ist eine Middleware, welche ein standartisiertes Interface für SQL Datenbanken bereit stellt.

Reviewer: ```suggestion
Die SmartAi kommuniziert mit dem Datenbankserver via JDBC, diese Kommunikation ist ebenfalls synchron. JDBC ist eine Middleware, welche ein standartisiertes Interface für SQL Datenbanken bereitstellt.
```

> Der Frontend Client kommuniziert mit dem SmartShopService mit HTTPS über eine REST-Schnittstelle. Die Kommunikation ist synchron.

Reviewer: ```suggestion
Der Frontend Client kommuniziert mit dem SmartShopService über HTTPS und eine REST-Schnittstelle. Die Kommunikation ist synchron.
```

> Der DatenbankService kommuniziert ähnlich wie die SmartAi über JDBC mit der Datenbank. Die Kommunikation ist synchron und JDBC ist eine Middleware.

Reviewer: ```suggestion
Der DatenbankService kommuniziert ähnlich wie die SmartAi über JDBC mit der Datenbank. Die Kommunikation ist synchron mit JDBC als Middleware.
```

> Der MqttService funktioniert nach dem Publish-Subscribe Prinzip. Der MqttService abonniert (subscribed) die Nachrichten vom MqttBroker, welcher wiederum die Nachrichten der Sensoren sammelt und an seine Abonnenten (Subscriber) weiter leitet.

Reviewer: ```suggestion
Der MqttService funktioniert nach dem Publish-Subscribe-Prinzip. Der MqttService subscribed sich auf die Nachrichten vom MqttBroker, welcher wiederum die Nachrichten der Sensoren sammelt und an seine Subscriber weiterleitet.
```

Der MqttBroker dient als Middleware, es wird das MQTT-Protokoll verwendet und die Kommunikation ist asynchron. \\

> Die nach außen bereitgestellten Schnittstellen sind vom Backend ausgehend REST-Schnittstellen, um Metainformationen anzufordern, wie zum Beispiel die aktuelle Auslastung eines Shops.

Reviewer: Der Satz ergibt so noch keinen Sinn, den musst du nochmal überarbeiten.

> Außerdem eine Schnittstelle die den besten Weg basierend auf beispielsweise dem aktuellen Standort, der bei der Anfrage mit geliefert wird, zurück gibt.

Reviewer: ```suggestion
Außerdem gibt es eine Schnittstelle die den besten Weg basierend auf dem aktuellen Standort und dem ausgewählten ShopType zurückgibt.
Beide Daten werden bei der Anfrage als Parameter mitgegeben.
```

Diese Schnittstellen werden im Regelfall vom Frontend angesprochen.

> Außerdem stellt der Datenbankserver eine Schnittstelle bereit, um mit SQL Daten anzufordern/zu speichern, allerdings wird diese nur vom Backend angesprochen und sollte von außen nur gesichert erreichbar sein. \\

Reviewer: ```suggestion
Außerdem stellt der Datenbankserver eine Schnittstelle bereit, um via SQL Daten anzufordern und zu speichern.
Dieser Endpunkt wird allerdings nur vom Backend über JDBC angesprochen und sollte von außen nicht erreichbar sein.
```

\subsection{Begründung}

> Die REST API wird in diesem Kontext verwendet, da sie eine einfache und standardisierte Kommunikation zwischen Services in einem verteilten System darstellt.

Reviewer: ```suggestion
Eine  REST API wird in diesem Kontext verwendet, da sie eine einfache und standardisierte Kommunikation zwischen Services in einem verteilten System ermöglicht.
```

> Dadurch ist gewährleistet, dass weitere Services einfach erweitert, ergänzt und ersetzt werden können, ohne die Schnittstelle selbst zu verändern.

Reviewer: ```suggestion
Dadurch ist gewährleistet, dass weitere Services einfach ergänzt und ersetzt beziehungsweise bestehende Services erweitert werden können, ohne dass die Schnittstelle selbst angepasst werden müssen.
```

Eine Alternative wäre es, Services direkt über die jeweilige IP anzusprechen.
Der Vorteil wäre, dass man dadurch eine größere Flexibilität und eine bessere Performance realisieren kann.
Der Nachteil dabei ist, dass der Programmierer die Kommunikation selbst implementieren muss und anpassen muss, sollte es zu Änderungen im System kommen.

> Außerdem bietet REST die Möglichkeit Services einfach nach außen, beispielsweise ins World Wide Web, verfügbar zu machen.

Reviewer: Das ist so nicht richtig. Der Aufruf über REST erfordert die jeweilige IP (bzw. einen Namen, der über DNS aufgelöst werden kann). Auch mit REST muss die Kommunikation vom Entwickler implementiert werden. Es gibt allerdings Frameworks, die hier einiges an Arbeit abnehmen.

Eine alternative zu REST wären beispielsweise Websockets, die beispielsweise als Basis für MessageBroker wie bei MQTT verwendet werden, allerdings ergibt es keinen Sinn, die bei dir zu verwenden. 

Diesen Absatz musst du nochmal überarbeiten.

> JDBC wird verwendet, da die Middleware ein standardisiertes Interface bietet und somit der Programmierer einfach mit dem Datenbank Server kommunizieren kann.

Reviewer: ```suggestion
JDBC wird verwendet, da es Middleware ein standardisiertes Interface bietet und somit der Programmierer einfach mit dem Datenbankserver kommunizieren kann.
```

> Ähnlich wie bei REST könnte man auch hier eine direkte Kommunikation implementieren, muss dann allerdings auch den Aufwand betreiben und bei Änderungen die Kommunikation anpassen.

Reviewer: ```suggestion
Ähnlich wie bei REST könnte man auch hier eine direkte Kommunikation implementieren, muss dann allerdings auch den Aufwand betreiben und bei Änderungen die Kommunikation anpassen.
Darüber hinaus muss im Datenbankserver eine andere Protokollform aktiviert werden, über die dann kommuniziert werden kann.
```

Der MqttBroker wird in diesem Kontext verwendet, da er die Kommunikation mit den Sensoren vereinfacht.

> So muss der Programmierer sich nicht darum kümmern wie viele Sensoren existieren und muss nur eingehende Änderungen verarbeiten.

Reviewer: ```suggestion
So muss der Programmierer sich nicht darum kümmern, wie viele Sensoren existieren und muss nur eingehende Änderungen verarbeiten.
Außerdem werden Änderungen in den Sensordaten so automatisch an den Server gesendet und der Server muss diese Änderungen nicht aktiv abfragen.
So wird Busy Waiting vermieden.
```

Zu dem Busy Waiting solltest du ggf. noch eine Quelle raussuchen.

> Hier wieder das Problem, man muss über eine Middleware arbeiten, welche nicht zwangsläufig optimiert funktioniert.

Reviewer: Der Satz ergibt so keinen Sinn.

\newpage

> \bibliography{lit}

Reviewer: ```suggestion
\bibliographystyle{alphadin}
\bibliography{lit}
```
Der Text kommt nicht hier hin, sondern in die Datei lit.bib. Dort verwendest du das .bib-Forrmat.

Tilkov, Stefan, et al. REST und HTTP: Entwicklung und 

> Integration nach dem Architekturstildes Web. dpunkt. verlag, 

Reviewer: ```suggestion
Integration nach dem Architekturstil des Web. dpunkt. verlag, 
```

2015

> \bibliographystyle{alpha}

Reviewer: ```suggestion
```

Durch die Kombination von statistischen Erfahrungswerten mit Umweltdaten wie dem Wetter, stattfindenden Events oder Baustellen und Streckensperrungen soll bestimmt werden, wann welcher Supermarkt am besten zu nutzen ist.
Darüber hinaus sollen Empfehlungen zur Route beziehungsweise den zu nutzenden öffentlichen Verkehrsmitteln gegeben werden.

Zur Anwendungsdomäne gehören folgende Entitäten:

\begin{description}
	\item[Sensordaten] Ein Sensor erfasst seinen aktuellen Standort, die aktuelle Zeit und die Anzahl an Kunden in seinem zugewiesenen Bereich. Ein solcher Sensor ist mittlerweile in vielen Supermärkten im Einsatz. Die oben genannten Attribute werden in der Entität \textit{SensorData} zusammengefasst und in regelmäßigen Abständen bereitgestellt.
	\item[Nutzer] Ein Nutzer hat einen Standort, welcher vom Client bei der Anfrage an das System weitergegeben wird. Zusätzlich wählt der Nutzer aus, nach welcher Art Shop er suchen möchte. Auf dieser Basis bekommt er einen Shop und gegebenenfalls eine Route empfohlen.
	\item[GeoPosition] Die GeoPosition gibt die geografische Position basierend auf dem Längen- und Breitengrad an.

> 	\item[SensorArea] Eine SensorArea ist ein abstraktes Konstrukt, das jeweils von Sensor und Shop implementiert wird. Er stellt einen Bereich dar der durch einen oder mehr Sensoren abgedeckt werden kann.

Reviewer 2: ```suggestion
	\item[SensorArea] Eine SensorArea ist ein abstraktes Konstrukt, das jeweils von Sensor und Shop implementiert wird. 
	Er stellt einen Bereich dar, der durch einen oder mehreren Sensoren abgedeckt werden kann.
```

	\item[Shop] Der Shop repräsentiert ein Geschäft oder eine Ansammlung von Geschäften, beispielsweise ein Einkaufszentrum. Zu einem Shop gehört jeweils ein Name, ein ShopType, die Größe der Ladenfläche, die Kapazität und die GeoPosition. Zusätzlich wird noch eine Liste an SensorAreas gespeichert.
	\item[ShopType] Der ShopType gibt an, um was für eine Art Shop es sich handelt.
Abhängig von der Art des Shops und seiner Größe wird auch die Kapazität berechnet.
	
\end{description}

\section{Architektur des Projektes}

\includegraphics[width=\linewidth]{images/deployment_diagram}

\subsection{Beschreibung der Systemarchitektur}

Der Server, auf dem die Anwendung läuft, kann beispielsweise ein Sun Ice 9001 sein.
Dieser wird mit dem Betriebssystem Ubuntu 20.04 LTS betrieben, da es Open Source, kostenlos und gut dokumentiert ist.
Darin wird zum einen die SmartAi.py ausgeführt, welche in der Laufzeitumgebung Python 3.5 läuft. 

> Python ist eine gängige Sprache für KI-Lösungen verwendet wird.

Reviewer 2: ```suggestion
Python ist eine gängige Sprache, die für KI-Lösungen verwendet wird.
```

Zum anderen wird dort die SmartShopping.jar in der Laufzeitumgebung Java 14 ausgeführt.
Auf Grund der im Studium gesammelten Erfahrung wurde sich für Java in Kombination mit Spring Boot entschieden.
Die Spring-Applikation stellt den in Spring Boot integrierten Embedded Tomcat bereit.
Die SmartAi.py und die SmartShopping.jar kommunizieren intern via REST-Schnittellen mit dem HTTP-Protokoll.
Mit dem Datenbankserver kommunizieren sie mit dem JDBC-Protokoll.
Der Datenbankserver wird beispielsweise auf einem IBM z15 ausgeliefert.
Auf diesem läuft eine MariaDB 10.6.4.
In der Datenbank gibt es die Schemata SensorData und Shop.

> Des Weiteren liefert das Backend den Frontend Client in Form einer Thymeleaf\footnote{\url{https://www.thymeleaf.org/}}-Anwendung aus.

Reviewer 2: Wenn du hier mit Fußnoten anfängst, dann fehlen welche auf alle anderen Technologien bisher. Die müsstest du eben noch setzen

Dieses Muster könnte durch das Zustands-Muster ausgetauscht werden.
Das Zustands-Muster würde in der konkreten Implementierung ebenfalls auf eingehende Nachrichten warten beziehungsweise reagieren, es könnte beispielsweise als ein Buffer agieren welcher Nachrichten sammelt und erst auf Anfrage Daten weiterleitet oder speichert.
Dadurch ist allerdings der Server für das stetige Abfragen der Daten verantwortlich.
Dieses sogenannte Busy Waiting ist sehr ressourcenintensiv und sollte daher vermieden werden (vgl. \cite{RJ2021}).

\newpage
\section{Beschreibung der Kernfunktionalität}
Eine der Kernfunktionalitäten der verteilten Anwendung ist es, die Auslastung von Shops zu ermitteln und mit diesen Informationen weiter zu arbeiten.
Daher stellen wir in diesem Abschnitt die Funktionalität zur Ermittlung der Auslastung eines Shops vom Typ ShoppingCenter dar.

\includegraphics[width=\linewidth]{images/Sequenzdiagramm}
In dem im Sequenzdiagramm beschriebenen Beispiel möchte ein Client die Auslastung eines bestimmten Shops abfragen und ruft daher die Methode calculateUsage auf dem SmartShopService auf.
\\

> \textbf{SmartShopService}: In diesem Beispiel wird davon ausgegangen, dass es bei der Auflösung einer bestimmten ShopOID nicht zu einem Fehler kommt sondern die Auflösung immer erfolgreich ist.

Reviewer 2: ```suggestion
\textbf{SmartShopService}: In diesem Beispiel wird davon ausgegangen, dass es bei der Auflösung einer bestimmten ShopOID nicht zu einem Fehler kommt, sondern die Auflösung immer erfolgreich ist.
```

\begin{lstlisting}[language=Java, basicstyle=\scriptsize]
public int calculateUsage(){
	return this.calculateCustomerCount() / this.capacity;
}
\end{lstlisting}

\newpage
\section{Persistente Datenhaltung}

\subsection{Persistierungansatz}
\includegraphics[width=\linewidth]{images/Datamodel}
\begin{description}
	\item[Shop] Der Shop besitzt eine eindeutige ID, welche als Primärschlüssel dient. 
	Des Weiteren besitzt er folgende Attribute: 
	\begin{description}
		\item [name:] Das Attribut name repräsentiert den Namen des Shops.
		\item [capacity:] Das Attribut capacity speichert die maximale Anzahl an Kunden, die sich in dem Shop aufhalten dürfen.
		\item [size:] Das Attribut size beschreibt die Grundfläche des Shops.
		\item [shopType:] ShopType repräsentiert den Typ des Shops.
		\item [lat, lng:] lat und lng beschreiben den Breiten- und Längengrad des Shops. Die Daten kommen aus der Entität \textbf{GeoPosition}, welche aber in diesem Kontext innerhalb des Shops persistiert wird, da Shop und GeoPosition in einer eins-zu-eins-Beziehung stehen.

> 		\item [parentOID:] Die parentOID ist ein Fremdschlüssel auf eine möglicherweise übergeordneten SensorArea, das Attribut kann leer sein.

Reviewer 2: ```suggestion
		\item [parentOID:] Die parentOID ist ein Fremdschlüssel auf eine möglicherweise übergeordneten SensorArea. Das Attribut kann leer sein.
```

	\end{description}
\end{description}

In der Datenhaltung werden die Entitäten Shop, Sensor und SensorData gespeichert.
Eine weitere Entität wäre die geografische Position, da diese aber in einer eins-zu-eins-Beziehung zur Entität Shop steht, werden die Daten direkt im Shop gespeichert.

Der DatenbankService verwendet Spring Data JPA.
Dadurch werden sowohl die benötigten Tabellen als auch die via Methodennamen definierten Datenbankabfragen automatisch generiert.
Die gewählte Datenbank ist eine MariaDB, welche auf einem eigenen Server ausgeliefert wird.

\subsection{Begründung}
Da die Daten einfach strukturiert sind und eine eindeutige ID besitzen, habe ich mich für eine SQL Datenbank entschieden.
Durch diese Datenstruktur ist ein schneller, sicherer und einfacher Zugriff möglich.
Die gewählte Datenbank ist eine MariaDB, da diese frei verfügbare OpenSource-Lösung zu MySQL ist.
Die Datenbank wird auf einem eigens dafür vorgesehenen Server aufgesetzt, um eine lose Koppelung zu gewährleisten. 
Außerdem können Anwendungen, anders als bei einer In-Memory Datenbank, unabhängig voneinander auf die Datenbank zugreifen.

> Ich habe mich für Spring Data JPA entschieden, da es durch Spring standardisiert ist und datenbankunabhängig funktioniert.

Reviewer 2: ```suggestion
Es wurde sich für Spring Data JPA entschieden, da es durch Spring standardisiert ist und datenbankunabhängig funktioniert.
```

> Außerdem nimmt es dem Programmierer einiges an Arbeit ab.

Reviewer 2: ```suggestion
Zudem unterstützt es den Entwickler bei seiner Arbeit  und verringert so den Aufwand von Routineaufgaben in der Datenbankanbindung.
```

Beispielsweise ist es nicht mehr notwendig eigene Abfragen zu definieren, da diese anhand der Methodennamen im entsprechenden Reopsitory Interface generiert werden.
Darüber hinaus werden die Tabellendefinitionen aus den Klassen generiert, sodass es nicht nötig ist, diese manuell zu erstellen.
Auch das Mapping von Fremdschlüsselbeziehungen auf die jeweiligen Objekt geschieht automatisch.
Diese Vorteile überwiegen den Nachteil, dass weitere Abhängigkeiten in den Code geladen werden.

Eine Alternative zu MariaDB wäre neben dem bereits erwähnten MySQL eine dokumentenorientierte Datenbank.

> Ich habe mich gegen diese Form von Datenbanken entschieden, da ich mit einem festen Schema arbeiten möchte und NoSQL dies nicht bietet.

Reviewer 2: ```suggestion
Es wurde sich gegen diese Form von Datenbanken entschieden, da mit einem festen Schema gearbeitet wird und NoSQL dies nicht unterstützt.
```

Graphdatenbanken eignen sich für diesen Anwendungsfall nicht, da die zu speichernden Daten nicht in einem Graph dargestellt werden können.
\newpage

\section{Kommunikation}

\subsection{Kommunikationsansatz}
\includegraphics[width=\linewidth]{images/Kommunikation}

Die Kommunikation zwischen SmartAI und dem AiService erfolgt über REST-Schnittstellen mit dem HTTP-Protokoll, diese Kommunikation ist synchron.

Die SmartAi kommuniziert mit dem Datenbankserver via JDBC, diese Kommunikation ist ebenfalls synchron. JDBC ist eine Middleware, welche ein standartisiertes Interface für SQL Datenbanken bereitstellt.

Der Frontend Client kommuniziert mit dem SmartShopService über HTTPS und eine REST-Schnittstelle. Die Kommunikation ist synchron.

Der DatenbankService kommuniziert ähnlich wie die SmartAi über JDBC mit der Datenbank. Die Kommunikation ist synchron mit JDBC als Middleware.

Der MqttService funktioniert nach dem Publish-Subscribe-Prinzip. Der MqttService subscribed sich auf die Nachrichten vom MqttBroker, welcher wiederum die Nachrichten der Sensoren sammelt und an seine Subscriber weiterleitet.

> Der MqttBroker dient als Middleware, es wird das MQTT-Protokoll verwendet und die Kommunikation ist asynchron. \\

Reviewer 2: ```suggestion
Der MqttBroker dient als Middleware, es wird das MQTT-Protokoll verwendet und die Kommunikation ist asynchron.
```

In diesem Kontext wurde sich gegen WebSocket entschieden, da die Kommunikation zwischen Client und Server synchron erfolgen kann.
Außerdem würden WebSockets eine nicht benötigte Komplexität in das Projekt bringen.

JDBC wird verwendet, da es als Middleware ein standardisiertes Interface bietet und somit der Programmierer einfach mit dem Datenbankserver kommunizieren kann.
Ähnlich wie bei REST kann auch hier eine direkte Kommunikation verwendet werden, es muss dann allerdings auch der Aufwand betreiben werden und bei Änderungen die Kommunikation anpassen.
Darüber hinaus muss im Datenbankserver eine andere Protokollform aktiviert werden, über die dann kommuniziert werden kann.

Der MqttBroker wird in diesem Kontext verwendet, da er die Kommunikation mit den Sensoren vereinfacht.
So muss der Programmierer sich nicht darum kümmern, wie viele Sensoren existieren und muss nur eingehende Änderungen verarbeiten.
Außerdem werden Änderungen in den Sensordaten so automatisch an den Server gesendet und der Server muss diese Änderungen nicht aktiv abfragen.
So wird Busy Waiting vermieden.

> Hier wieder das Problem, man muss über eine Middleware arbeiten, welche nicht zwangsläufig optimiert funktioniert.

Reviewer 2: ```suggestion
Hier existiert wieder das Problem, dass Middleware eingesetzt werden muss, welche nicht zwangsläufig für diesen Anwendungsfall optimiert ist.
```
